=clu #18554 Make oldest assignment deterministic when joining

patriknw opened the pull request:
- the reported issue is fixed by the immediate leaderActions
  (moving to Up)  when joining the first node to itself
- the other changes are precautions just in case

patriknw committed:
=clu #18554 Make oldest assignment deterministic when joining

* the reported issue is fixed by the immediate leaderActions
  (moving to Up)  when joining the first node to itself
* the other changes are precautions just in case

patriknw commented:
Refs #18554

ktoso commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala:
```diff
@@ -325,7 +325,7 @@ class ShardRegion(
     case MemberUp(m) ⇒
       if (matchingRole(m))
-        changeMembers(membersByAge + m)
+        changeMembers(membersByAge - m + m) // replace
```
thanks for the comment, for a second thought was confused ;)

ktoso commented on akka-cluster-tools/src/test/java/akka/cluster/singleton/ClusterSingletonManagerTest.java:
```diff
@@ -46,77 +46,6 @@ public void demo() {
   }
 
-  static//documentation of how to keep track of the oldest member in user land
-  //#singleton-proxy
```
oh seems it wasn't ever linked to from docs. OK

ktoso commented:
LGTM.

Would it make sense to mention the `first joining itself will immediately be moved to Up` in the docs at least with a few words?

patriknw commented:
I think that is an implementation detail, which no happen to coincide with what users always have expected (first node joining is also the oldest and it will of course be moved to Up asap)

ktoso commented:
True, good as is then :)

viktorklang commented on akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonProxy.scala:
```diff
@@ -203,8 +201,9 @@ final class ClusterSingletonProxy(singletonManagerPath: String, settings: Cluste
-      trackChange {
-        () ⇒ membersByAge += m
+      trackChange { () ⇒
+        membersByAge -= m // replace
```
For consistency?

~~~scala
membersByAge = membersByAge - m + m // replace
~~~

viktorklang commented on akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala:
```diff
@@ -275,7 +275,8 @@ object ClusterEvent {
+        case (_, newMember :: oldMember :: Nil) if newMember.status != oldMember.status || newMember.upNumber != oldMember.upNumber ⇒
+          newMember
       }
       val memberEvents = (newMembers ++ changedMembers) collect {
```
Hmm, didn't that performance article recommend using `diff` and `union` iso -- and ++?

patriknw replied:
@viktorklang yes, that was something I learned from that presentation (surprised me that they are not implemented in the same way). I have changed: https://github.com/akka/akka/commit/811e28595f24de89210ceb85684c93fdb75ce0cb

viktorklang replied:
:+1:

patriknw committed:
=clu replace Set -- with diff and ++ with union

* better performance according to
  https://docs.google.com/presentation/d/1Qjryxoe-fYEM8ZPhM-98LKfbhnRcn5eAEMNlVVnixsA/pub
